### ISSUE TITLE
Minor bug in FTP response codes

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4702]
|<img alt="JohnDoeee's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| JohnDoeee reported|
|-|-|
|Trac ID|trac#4702|
|Type|defect|
|Created|2010-10-19 09:07:01Z|
|Branch|https://github.com/twisted/twisted/tree/ftp-response-code-4702|

There is a minor bug in the FTP response codes making it impossible to ever use the response text for CLOSING_DATA_CNX.

Attached patch should solve this problem.

Attachments:

* [ftp-bugfix.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/49b/49b0750be52d50241875edd3423fab8a36668398/e67248539001c1f36d1afcb8970272f1179ff51c.diff) (596 bytes) - added by JohnDoeee on 2010-10-19 09:07:16Z - 
* [ftp-bugfix-v2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/49b/49b0750be52d50241875edd3423fab8a36668398/67475275966dad89fda080eb013516ece7285281.diff) (1997 bytes) - added by JohnDoeee on 2010-11-01 16:27:13Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4702 4702
type__defect defect
reporter__JohnDoeee JohnDoeee
priority__low low
milestone__ 
branch__branches_ftp_response_code_4702 branches/ftp-response-code-4702
branch_author__JohnDoeee__exarkun JohnDoeee, exarkun
status__closed closed
resolution__fixed fixed
component__ftp ftp
keywords__ 
time__1287479221000000 1287479221000000
changetime__1302448677000000 1302448677000000
version__None None
owner__exarkun exarkun
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_2"></a>|@jesstess set owner to JohnDoeee|
|-|-|

Thanks for the bug report and patch, JohnDoeee! Can you add unit tests for your changes as part of this patch, so this functionality won't break unnoticed again? Some documentation on Twisted's testing policy can be found [here](http://twistedmatrix.com/documents/current/core/development/policy/test-standard.html); I'm happy to help if you have any questions about writing the tests.

**Comment by twisted-trac** [comment]
|<img alt="JohnDoeee's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|JohnDoeee commented|
|-|-|

The test should check for duplicate use, but it seems like the duplicate codes are actually not used anywhere.

Anyways, should not be left as that, the patch will make sure it won't happen again.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

Reporter attached a patch with unit tests, needs review now.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [31570]) Branching to 'ftp-response-code-4702'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

(In [31571]) Apply ftp-bugfix-v2.diff

refs [#4702](https://github.com/twisted/twisted/issues/4702)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  Sorry for the delay on the review; make sure to always add the _review_ keyword when you're ready for someone to take a look. :)

  1. `FTPResponseCodeTest` should document what the collection of test methods it will define is for; even if it only has one test method.  The individual test method should have a docstring explaining a specific feature being tested.
  1. Test method docstrings should be written in an imperative style.  "There are no duplicate FTP response codes."  It's a test method, so _Test to check ..._ is redundant.
  1. `isinstance` should be preferred over directly `type(foo) == bar` checks
  1. trial has `assertNotIn` which is nicer than `assertFalse(x in y)`

I'm going to take a crack at addressing these points and hopefully we can put this to rest.  Thanks again for the contribution!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [31572]) Coding standard fixes and strengthen the assertion made by the test: response code values must be unique and all response code constants must be present in the RESPONSE dict

refs [#4702](https://github.com/twisted/twisted/issues/4702)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun removed owner|
|-|-|

Made the changes above, plus a bit more.  Some of the response code globals were missing from the `RESPONSE` dict.  The test verifies they are all in it now, and I added the ones which were missing.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/ftp-response-code-4702)

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_12"></a>|@jesstess set owner to @exarkun|
|-|-|

Thanks for seeing this through, exarkun.

There are two comments in ```ftp.py``` that should have tickets:

```    RESTART_MARKER_REPLY:               '110 MARK yyyy-mmmm', # TODO: this must be fixed```

and

```    'userotp':                          '331 Response to %s.',  # ???```

Other than that, the branch looks good, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Thanks for the review.  I filed [#5045](https://github.com/twisted/twisted/issues/5045) and [#5046](https://github.com/twisted/twisted/issues/5046) for those two issues.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set status to `closed`|
|-|-|

(In [31598]) Merge ftp-response-code-4702 

Author: JohnDoeee, exarkun
Reviewer: jesstess
Fixes: [#4702](https://github.com/twisted/twisted/issues/4702)

Assign unique values to each FTP response code constant so that each
can have a corresponding response message assigned to it.
